clickstream/src/main/kotlin/clickstream/ClickStream.kt

```diff
@@ -39,6 +40,22 @@ public interface ClickStream {
      */
     public fun trackEvent(event: CSEvent, expedited: Boolean)
 
+    /**
+     * Configure a single client scoped listener that will receive all analytic events
+     * for this client.
+     *
+     * @see CSEventListener for semantics and restrictions on listener implementations.
+     */
+    public fun addEventListener(listener: CSEventListener)
+
+    /**
+     * removes the listeners added by [addEventListener].
+     *
+     * @see CSEventListener for semantics and restrictions on listener implementations.
+     */
+    public fun removeEventListener(listener: CSEventListener)
```

Comment:
Maybe I'm missing something here but curious: When do we foresee the client would ever need to remove a listener?

I guess for things like UI, this kind of API makes sense (required, even) because their lifecycle tends to be rather short. Clickstream, on the other hand, will probably continue to live as long as its host application does. Like the visualizer, for example. It's there and we want it to remain there for the length of Clickstream's lifecycle.

I understand the argument above contains a lot of assumptions. So please do CMIIW.

Reply:
Hey @MrHadiSatrio,

You are right, the current implementation for ev will only require the listener to be added once. However, the reason for adding a `removeEventListener ` are :
 
• Sake of having a complete API.
• Let's say someone wants to just listen for events in an activity. He can `addEventListener` in onCreate and `removeListener` onDestroy. This would save memory as currently, we are maintaining a list of listeners in the clickstream.

Reply:
@kshitij6325 shouldn't we remove the listener whenever we toggled off EV Ui?, by doing this we're avoiding unnecessary overhead doing computation of load the event, sent to Memcache, also upstream callbacks